assembly/src/assembler/instruction/crypto_ops.rs

```diff
@@ -96,19 +96,58 @@ pub(super) fn mtree_get(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, Ass
 ///
 /// This operation takes 14 VM cycles.
 pub(super) fn mtree_set(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, AssemblyError> {
+    // input: [d, i, R_old, V_new, ...]
+
     // Inject the old node value onto the stack for the call to MRUPDATE.
-    // [d, i, R, V_new, ...] => [V_old, d, i, R, V_new, ...]
+    // stack: [V_old, d, i, R_old, V_new, ...]
     read_mtree_node(span);
+
+    // Note: The stack is 14 elements deep already. The existing ops manipulate up to depth 16, so
+    // it's only possible to copy 2-elements at the time.
+    #[rustfmt::skip]
+    let copy_old_element_ops = [
+        // Renamed V_old => o<pos>, R_old => r<pos>, V_new => n<pos>
+        // stack: [[o3, o2, o1, o0], [d, i, r3, r2], [r1, r0, n3, n2], n1, n0, ...]
+
+        // Copy last two elements of the word, and move it down
+        // stack: [[o3, o2, o1, o0], [d, i, r3, r2], [o0, o1, r1, r0], [n3, n2, n1, n0], ...]
+        Dup2, MovDn8, Dup3, MovDn8,
+
+        // Fix the position of the new elements
+        // stack: [[o0, o1, r1, r0], [n3, n2, n1, n0], [o3, o2, o1, o0], [d, i, r3, r2], ...]
+        SwapDW,
+
+        // stack: [[r1, r0, n3, n2], [n1, n0, o1, o0], [o3, o2, o1, o0], [d, i, r3, r2], ...]
+        MovDn7, MovDn6,
+
+        // The stack is restored, but with the two elements copied at the end
+        // stack: [[o3, o2, o1, o0], [d, i, r3, r2], [r1, r0, n3, n2], [n1, n0, o1, o0], ...]
+        SwapDW,
+
+        // Repeat for the other two elements
+        Dup1, MovDn8, Dup0, MovDn8, SwapDW, MovDn7, MovDn6, SwapDW
+    ];
```

Comment:
I think maybe we could do this in fewer cycles via the following

```
        MovUp5,
        MovUp5,
        Dup5,
        Dup5,
        SwapDW,
        SwapW, SwapW2,
        Dup13,
        Dup13
 ```
 
 An interesting note here is that the costs of the `Dup.n` and `MovUp.n`/`MovDn.n` operations work differently. For the `Mov` ops, it's one cycle for each op through index `8`, then 4 cycles thereafter. For the `Dup` ops, it's 1 cycle up through index `7` and for every dup of an odd index. Obviously all of this will be easier to work with once we have a better way of looking up cycle costs of operations, since the current docs just give ranges.

This is a good place to check for now (and you may have already looked there): https://github.com/0xPolygonMiden/miden-vm/blob/main/assembly/src/assembler/instruction/mod.rs

Reply:
Uh, neat, so instead of making a new copy and pushing it down like I did. You pushed down and copied up. Thanks!

Reply:
> An interesting note here is that the costs of the Dup.n and MovUp.n/MovDn.n operations work differently. For the Mov ops, it's one cycle for each op through index 8, then 4 cycles thereafter

Thanks for the reference! I had seen that because the span accepts Operations instead Instruction, so I was using the lower level ops already. One question though, is it correct to say that every op takes a single cycle? Or are there any ops that takes more than one?